server/log_operation_manager.go

```diff
@@ -54,27 +62,142 @@ type LogOperationInfo struct {
 	// batch takes longer than this interval to complete, the next batch
 	// will start immediately.
 	RunInterval time.Duration
+	// PreElectionPause is the maximum interval to wait before starting a
+	// mastership election for a particular log.
+	PreElectionPause time.Duration
+	// MasterCheckInterval is the interval between checks that we still
+	// hold mastership for a log.
+	MasterCheckInterval time.Duration
+	// MasterHoldInterval is the minimum interval to hold mastership for.
+	MasterHoldInterval time.Duration
+	// ResignOdds gives the chance of resigning mastership after each
+	// check interval, as the N for 1-in-N.
+	ResignOdds int
 	// NumWorkers is the number of worker goroutines to run in parallel.
 	NumWorkers int
 }
 
+type electionRunner struct {
+	logID    int64
+	info     *LogOperationInfo
+	tracker  *util.MasterTracker
+	cancel   context.CancelFunc
+	wg       *sync.WaitGroup
+	election util.MasterElection
+}
+
+func (er *electionRunner) Run(ctx context.Context) {
+	defer er.wg.Done()
+
+	// Pause for a random interval so that if multiple instances start at the same
+	// time there is less of a thundering herd.
+	maxPause := er.info.PreElectionPause
+	if maxPause < minPreElectionPause {
+		maxPause = minPreElectionPause
+	}
+	pause := rand.Int63n(maxPause.Nanoseconds())
+	time.Sleep(time.Duration(pause))
+
+	glog.V(1).Infof("%d: start election-monitoring loop ", er.logID)
+	if err := er.election.Start(ctx); err != nil {
+		glog.Errorf("%d: election.Start() failed: %v", er.logID, err)
+		return
+	}
+	defer func(ctx context.Context, er *electionRunner) {
+		glog.Infof("%d: shutdown election-monitoring loop", er.logID)
+		er.election.Close(ctx)
+	}(ctx, er)
+
+	for {
+		glog.V(1).Infof("%d: When I left you, I was but the learner", er.logID)
+		if err := er.election.WaitForMastership(ctx); err != nil {
+			glog.Errorf("%d: er.election.WaitForMastership() failed: %v", er.logID, err)
+			return
+		}
+		glog.V(1).Infof("%d: Now, I am the master", er.logID)
+		er.tracker.Set(er.logID, true)
+		masterSince := time.Now()
+
+		// While-master loop
+		checkInterval := er.info.MasterCheckInterval
```

Comment:
you could do this clamp-upwards checks up top next to the pre-election one (and possibly move them out into a fixMyInputs type func which gets called at the top of this func